src/vs/workbench/parts/debug/electron-browser/debugHover.ts

```diff
@@ -112,27 +112,75 @@ export class DebugHoverWidget implements editorbrowser.IContentWidget {
 			return;
 		}
 
-		// string magic to get the parents of the variable (a and b for a.b.foo)
+		const session = this.debugService.getActiveSession();
+		const canEvaluateForHovers: boolean = session.configuration.capabilities.supportsEvaluateForHovers;
+
 		const lineContent = model.getLineContent(pos.lineNumber);
-		const namesToFind = lineContent.substring(0, lineContent.indexOf('.' + hoveringOver))
-			.split('.').map(word => word.trim()).filter(word => !!word);
-		namesToFind.push(hoveringOver);
-		namesToFind[0] = namesToFind[0].substring(namesToFind[0].lastIndexOf(' ') + 1);
+		let evaluatedExpression : TPromise<Expression> = undefined;
+		let matchingExpression: string = undefined;
+		let startOffset: number = 0;
+		if (canEvaluateForHovers) {
+			// If the debug adapter supports evaluation-based-hover, we need to try and guess what the expression under the mouse cursor is.
+			// Someday it might be nice if the language service could somehow provide this.
+			// But this code attempts to approximate the answer for a variety of languages.
+			// Some example supported expressions: myVar.prop, a.b.c.d, myVar?.prop, myVar->prop, MyClass::StaticProp, *myVar
+
+			// Match any character except a set of characters which often break interesting sub-expressions
+			let expression: RegExp = /([^()\[\]{}<>\s+\-/%~#^;=|,`!]|\->)+/g;
```

Comment:
ideally 'debugHover' should be implemented per debugger extension, with the help of language service. this is a stop-gap solution which should work for most cases. it can be improved when the necessary api is available.